### PR TITLE
network: support UID based routing policy

### LINKED ISSUE #14666: networkd RoutingPolicyRule uidrange support
**Is your feature request related to a problem? Please describe.**
I need an ability to add rules like this: `ip rule add uidrange 123-123 table custom`

**Describe the solution you'd like**
Add a `UIDRange` option to the `[RoutingPolicyRule]` section of `*.network` files.

### PR DESCRIPTION
**Pull request by yuwata**
Closes #14666.

### CODE REVIEW

**Comment by yuwata** [comment]
cc @ssahani.

File: src/basic/user-util.c [hunk 1]
```diff
@@ -62,6 +62,29 @@ int parse_uid(const char *s, uid_t *ret) {
...
+        r = parse_range(s, &l, &u);
+        if (r < 0)
+                return r;
+
+        if (l > u)
+                return -EINVAL;
+
+        if (!uid_is_valid(l) || !uid_is_valid(u))
+                return -ENXIO;
+
```
Comment (ssahani) [inline src/basic/user-util.c:82]: ```return -EINVAL;```

File: src/basic/user-util.c [hunk 2]
```diff
@@ -62,6 +62,29 @@ int parse_uid(const char *s, uid_t *ret) {
...
+        *ret_lower = l;
+        *ret_upper = u;
+        return 0;
+}
+
 char* getlogname_malloc(void) {
         uid_t uid;
```
Comment (ssahani) [inline src/basic/user-util.c:89]: `char *getlogname_malloc`

File: src/basic/user-util.c (hunk 1 above)
Comment (yuwata) [reply src/basic/user-util.c:82]: Please see the comment in `parse_uid()`. This is intended.

File: src/basic/user-util.c (hunk 2 above)
Comment (yuwata) [reply src/basic/user-util.c:89]: Yeah, I prefer `char *`. But `char* ` vs `char *` is out of scope in this PR.

**Comment by ssahani** [comment]
CI is failing and it's networkd test case . Please return the CI

File: src/basic/user-util.c (hunk 1 above)
Comment (ssahani) [reply src/basic/user-util.c:82]: yes right

**Comment by ssahani** [comment]
LGTM We can merge one CI pass

File: src/network/networkd-routing-policy-rule.c [hunk 3]
```diff
@@ -1170,6 +1230,13 @@ int routing_policy_serialize_rules(Set *rules, FILE *f) {
                         space = true;
                 }
 
+                if (rule->uid_range.start != UID_INVALID && rule->uid_range.end != UID_INVALID) {
+                        fprintf(f, "%suidrange=%"PRIu32"-%"PRIu32,
```
Comment (poettering) [inline src/network/networkd-routing-policy-rule.c]: UID_FMT instead of "%" PRIu32 please!

**Comment by poettering** [comment]
lgtm, only that nitpick

File: src/network/networkd-routing-policy-rule.c (hunk 3 above)
Comment (yuwata) [reply src/network/networkd-routing-policy-rule.c]: Hmm, maybe no, as `uid_range.start` or `.end` are `uint32_t`, not `uid_t`. `uid_t` may be `uint16_t`.

Comment (yuwata) [reply src/network/networkd-routing-policy-rule.c]: I noticed that we have `assert_cc(sizeof(uid_t) == sizeof(uint32_t));` at several places. So, I've replaced `"%"PRIu32` with `UID_FMT`.

**Comment by eternal-sorrow** [comment]
Just tested this on my machine, worked like a charm!

**Comment by yuwata** [comment]
Let's merge this. As this already got positive review by @poettering. And the change since is trivial.

**Pull request merged**